pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/ReferralSourceService.java

```diff
@@ -0,0 +1,408 @@
+
+        try {
+            Map<String, String> props = getDataMap(participantData);
+            if (props.isEmpty()) {
+                throw new DsmInternalError("Participant data empty " + msg);
+            }
+
+            // do not overwrite an existing referral source, unless marked as "NA" which is a placeholder
+            if (!isProband(props) || (props.containsKey(DBConstants.REFERRAL_SOURCE_ID)
+                    && !props.get(DBConstants.REFERRAL_SOURCE_ID).equals(NA_REF_SOURCE))) {
+                return false;
+            }
+            log.info("Updating REFERRAL_SOURCE data {}", msg);
+            props.put(DBConstants.REFERRAL_SOURCE_ID, refSourceId);
+            ParticipantDataDao dataDao = new ParticipantDataDao();
+            dataDao.updateParticipantDataColumn(
+                    new ParticipantData.Builder()
+                            .withParticipantDataId(participantData.getParticipantDataId())
+                            .withDdpParticipantId(ddpParticipantId)
+                            .withDdpInstanceId(participantData.getDdpInstanceId())
+                            .withFieldTypeId(RGP_PARTICIPANT_DATA)
+                            .withData(gson.toJson(props))
+                            .withLastChanged(System.currentTimeMillis())
+                            .withChangedBy(userId).build());
+        } catch (JsonSyntaxException | JsonIOException | ClassCastException e) {
+            throw new DsmInternalError("Invalid data format " + msg, e);
+        }
+        return true;
+    }
+
+    protected static Map<String, String> getDataMap(ParticipantData participantData) throws JsonSyntaxException {
+        Optional<String> data = participantData.getData();
+        if (data.isEmpty() || StringUtils.isEmpty(data.get())) {
+            return Collections.emptyMap();
+        }
+
+        return gson.fromJson(data.get(), Map.class);
+    }
+
+    private List<Activities> getParticipantActivities(String ddpParticipantId, String esIndex) {
+        Optional<ElasticSearchParticipantDto> esParticipant =
+                ElasticSearchUtil.getParticipantESDataByParticipantId(esIndex, ddpParticipantId);
+        if (esParticipant.isEmpty()) {
+            throw new ESMissingParticipantDataException("Participant ES data is null for participant " + ddpParticipantId);
+        }
+        return esParticipant.get().getActivities();
+    }
+
+    /**
+     * Derive REFERRAL_SOURCE ID from participant activities
+     *
+     * @param activities list of DSS participant activities
+     * @return a referral source ID
+     * @throws DsmInternalError If the expected referral source values are not present or the referral source mapping is
+     *                          missing or out of sync
+     */
+    public static String deriveReferralSourceId(List<Activities> activities) {
+        return ReferralSourceService.convertReferralSources(ReferralSourceService.getReferralSources(activities));
+    }
+
+    /**
```

Comment:
Note to reviewer: the following three methods were reviewed in May, when they were part of a PR to support PEPPER-575.